### PR TITLE
markdown printer

### PR DESCRIPTION
**Pull request by GantMan**
Take this data
```js
    const data = [
      ['Header 1', 'Header 2', 'Header 3'],
      ['matthew1', '1', 'howdy1'],
      ['matthew2', '2', 'howdy2'],
      ['matthew3', '3', 'howdy3'],
      ['matthew', '4', 'howdy4']
    ]
```
then do this:
`context.print.table(data, {markdown: true})`

it will output this:
image

When pasted in github it will do this!

| Header 1 | Header 2 | Header 3 |
| ---      | ---      | ---      |
| matthew1 | 1        | howdy1   |
| matthew2 | 2        | howdy2   |
| matthew3 | 3        | howdy3   |
| matthew  | 4        | howdy4   |

### CODE REVIEW

**Comment by GantMan** [comment]
No tests are written yet.  I wish that I didn't have to inject `---` but could have a full sized spacer on the header.  This hack works though and looks nice on CLI output.  Just not ideal.

I can add tests if we are happy.

**Comment by jamonholmgren** [comment]
One thing about this is I'm not entirely happy with this being outside of an extension. (not your fault, @GantMan , it was here before). I'd like to reconsider where _all_ of the files in `utils` should live.

I suppose we don't need to discuss that for the purposes of this PR, though.

**Comment by GantMan** [comment]
@jamonholmgren - I def like the format being an enumerated list of strings.  Good call.

`format = 'markdown' | 'lean' | 'default'` etc.

Jamon any ideas on how to do those dashes so they fit column length?  I can't find a way to do optional separators in `cli-table2` so I'm injecting 3 dashes instead.

**Comment by jamonholmgren** [comment]
Got it figured out. Commit incoming.

```javascript
    context.print.table(
      [
        ['header 1', 'header 2', 'header 3'],
        ['12', 'Four score and seven years ago', '94r5511244'],
        ['12', '99 bottles', '94r5511244'],
      ],
      { format: 'markdown' }
    )
```

image

**Comment by skellock** [comment]
@GantMan It looks like they don't support that option for customizing that header bottom border.

Bummer dude.

**Comment by skellock** [comment]
Classic @jamonholmgren , figuring it out while i'm the process of saying it's not supported.

File: docs/context-print.md [hunk 1]
```diff
@@ -1,50 +1,54 @@
 Features for allowing you to print to the console.
 
-
```
Comment (jamonholmgren) [inline docs/context-print.md]: Sorry about formatting noise. I'll comment where you need to look.

File: docs/context-print.md [hunk 2]
```diff
@@ -124,3 +131,31 @@ You can pass in a "command path" to refine what commands you'd like to see:
 const { printCommands } = context.print
 printCommands(context, ['generate', 'model'])
 ```
+
+## table
```
Comment (jamonholmgren) [inline docs/context-print.md]: New doc for `table`.

File: src/cli/commands/generate/generate.js [hunk 3]
```diff
@@ -1,8 +0,0 @@
-module.exports = {
```
Comment (jamonholmgren) [inline src/cli/commands/generate/generate.js]: Removed some CLI commands -- ignore this, I'll be doing more work in here later too.

File: src/index.js [hunk 4]
```diff
@@ -20,6 +20,11 @@ if (!sniff.isNewEnough) {
   process.exit(1)
 }
 
+// we want to see real exceptions with backtraces and stuff
+process.on('unhandledRejection', up => {
+  throw up
+})
```
Comment (jamonholmgren) [inline src/index.js]: @skellock 's magic for getting better backtraces.

File: src/utils/print.js [hunk 5]
```diff
@@ -2,6 +2,31 @@ const colors = require('colors')
...
+  'right-mid': '',
+  middle: ' '
+}
+
+const CLI_TABLE_MARKDOWN = {
+  ...CLI_TABLE_COMPACT,
+  left: '|',
+  right: '|',
+  middle: '|'
+}
```
Comment (jamonholmgren) [inline src/utils/print.js]: @GantMan I moved the table formatting to constants.

File: src/utils/print.js [hunk 6]
```diff
@@ -29,33 +54,51 @@ function divider () {
...
+    []
+  )
+}
+
+/**
+ * Returns an array of column headers based on column widths.
+ */
+function columnHeaderDivider (table) {
+  return findWidths(table).map(w => Array(w).join('-'))
+}
```
Comment (jamonholmgren) [inline src/utils/print.js]: These two functions give us the row we need to add, `['----', '--', '---------']` etc

File: src/utils/print.js [hunk 7]
```diff
@@ -29,33 +54,51 @@ function divider () {
...
-      mid: '',
-      'mid-mid': '',
-      right: '',
-      'right-mid': '',
-      middle: ' '
-    }
-  })
-  t.push(...data)
+  let t
+  switch (options && options.format) {
```
Comment (jamonholmgren) [inline src/utils/print.js]: Changed to a `switch case` statement instead of if/else.

File: src/utils/print.test.js [hunk 8]
```diff
@@ -82,6 +82,12 @@ test.serial('table', t => {
   t.is(stripAnsi(spyLog.args[i++][0]), '  liam      5 \n  matthew   2 ')
 })
 
+test.serial('markdown table', t => {
+  const data = [['liam', '5'], ['matthew', '2']]
+  print.table(data, { format: 'markdown' })
+  t.is(stripAnsi(spyLog.args[i++][0]), '| liam    | 5 |\n| ------- | - |\n| matthew | 2 |')
+})
```
Comment (jamonholmgren) [inline src/utils/print.test.js]: Added a test.

**Comment by jamonholmgren** [comment]
scored

File: src/index.js (hunk 4 above)
Comment (GantMan) [reply src/index.js]: lol throw up
🤢

File: src/utils/print.js (hunk 5 above)
Comment (GantMan) [reply src/utils/print.js]: Love it.  I was fighting the good fight so hard, I didn't clean anything.   Sorry to leave my PR looking like Freshman dorm.

File: src/utils/print.js [hunk 9]
```diff
@@ -29,33 +54,51 @@ function divider () {
...
 
+/**
+ * Returns an array of the column widths.
+ */
+function findWidths (table) {
+  return [table.options.head, ...table].reduce(
+    (colWidths, row) => row.map((str, i) => Math.max(`${str}`.length + 1, colWidths[i] || 1)),
+    []
+  )
+}
```
Comment (GantMan) [inline src/utils/print.js]: woah!

**Comment by GantMan** [comment]
I was close
close

**Comment by GantMan** [comment]
@jamonholmgren - I think you should be the one to hit merge.

Can I get a beta version shipped with this feature in it?  I want to use our new table in Solidarity.

**Pull request merged**